### ISSUE TITLE
db: Transaction with deletion then insertion of same record fails

### ISSUE THREAD

**Issue by fabioberger** [0xProject/0x-mesh#565]
Attempting to delete and then re-insert the same record within a DB transaction fails:

Example:
```
txn := d.meshDB.MiniHeaders.OpenTransaction()
defer func() {
	_ = txn.Discard()
}()

// Clear existing, then insert updated
var storedHeaders []*miniheader.MiniHeader
if err := d.meshDB.MiniHeaders.FindAll(&storedHeaders); err != nil {
	return err
}
for _, header := range storedHeaders {
	if err := txn.Delete(header.ID()); err != nil {
		return err
	}
}
for _, miniHeader := range d.miniHeaders {
	if err := txn.Insert(miniHeader); err != nil {
		return err
	}
}

if err := txn.Commit(); err != nil {
	return err
}
```

fails at the insert with error: `model already exists with the given ID: XXX` but should be possible.

**Comment by albrow** [comment]
I spent a little bit of time looking into this. Unfortunately there is no way to work around it without giving up some guarantees about consistency and atomicity. The problem occurs because no updates within a transaction actually have an effect until after the transaction is committed. When you call `Insert` or `Delete`, we need to check the current state of the database (which doesn't include any state changes within the transaction).

What we can do is return a better error message and make transactions more strict (currently calling `Insert` or `Update` twice for the same model within a single transaction can cause problems, but doesn't return an error).

**Comment by albrow** [comment]
This was addressed in #589.
